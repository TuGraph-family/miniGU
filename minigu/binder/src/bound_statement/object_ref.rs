use gql_parser::ast::{GraphElementType, Ident, SchemaPath, SchemaPathSegment};
use macro_rules_attribute::apply;
use serde::Serialize;
use minigu_catalog::types::SchemaId;
use crate::bound_statement::catalog::BoundGraphTypeRef;
use crate::bound_statement::expr::BoundGraphExpr;
use crate::catalog_ref::{CallProcedureCatalogRef, GraphTypeCatalogRef};
use crate::macros::base;


// Standard schema path, without relative paths and variables.
#[apply(base)]
pub struct CanonicalSchemaPath {
    pub segments: Vec<Ident>,
}


#[derive(Debug, Serialize)]
pub enum BoundProcedureRef {
    // Catalog ref or just parameter.
    Ref(CallProcedureCatalogRef),
    Parameter(Ident),
}
#[derive(Debug, Serialize)]
pub enum BoundOfGraphType {
    Like(BoundGraphExpr),
    Ref(BoundGraphTypeRef),
    Nested(Vec<GraphElementType>),
    Any,
}

#[derive(Debug, Serialize)]
pub enum BoundGraphTypeSource {
    Copy(BoundGraphTypeRef),
    Like(BoundGraphExpr),
    Nested(Vec<GraphElementType>),
}

#[apply(base)]
pub enum CatalogObjRefType {
    // To identify the category to which the ID in the catalog object belongs
    GraphRef,
    ProcedureRef,
    GraphTypeRef,
    BindingTableRef,
}

impl CanonicalSchemaPath {
    pub fn to_string(&self) -> String {
        format!("/{}", self.segments.join("/"))
    }

    pub fn default() -> Self {
        CanonicalSchemaPath { segments: vec![] }
    }
}

pub fn normalize_path(path: &SchemaPath) -> CanonicalSchemaPath {
    let mut stack = Vec::new();
    for segment in path.iter() {
        match &segment.value() {
            SchemaPathSegment::Name(ident) => {
                stack.push(ident.clone());
            }
            // If empty, ignore it. This indicates that the root directory was reached by ascending
            // via "..".
            SchemaPathSegment::Parent => {
                stack.pop();
            }
        }
    }
    CanonicalSchemaPath { segments: stack }
}