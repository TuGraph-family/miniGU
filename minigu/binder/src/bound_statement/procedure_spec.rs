use macro_rules_attribute::apply;
use serde::Serialize;
use gql_parser::ast::{BindingVariableDefBlock, Yield};
use gql_parser::span::OptSpanned;
use crate::bound_statement::catalog::LinearBoundCatalogModifyingStatement;
use crate::bound_statement::data::BoundLinearDataModifyingStatement;
use crate::bound_statement::query::BoundCompositeQueryStatement;
use crate::catalog_ref::SchemaCatalogRef;
use crate::macros::base;

#[derive(Debug, Serialize)]
pub struct BoundProcedure {
    pub at: Option<SchemaCatalogRef>,
    pub binding_variable_def: BindingVariableDefBlock,
    pub statement: BoundStatement,
    pub next_statement: Vec<BoundNextStatement>,
}


#[derive(Debug, Serialize)]
pub enum BoundStatement {
    Catalog(LinearBoundCatalogModifyingStatement),
    Query(BoundCompositeQueryStatement),
    Data(BoundLinearDataModifyingStatement),
}


#[derive(Debug, Serialize)]
pub struct BoundNextStatement {
    pub yield_clause: OptSpanned<Yield>,
    pub statement: BoundStatement,
}
