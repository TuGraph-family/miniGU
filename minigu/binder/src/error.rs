use std::fmt::{Debug, Display, Formatter};
use std::marker::PhantomData;
use std::ops::Range;
use std::sync::Arc;
use std::result;
use gql_parser::error::TokenError;
use crate::{impl_binder_already_exist_error_display, impl_binder_not_exist_error_display};


#[cfg(feature = "miette")]
use miette::Diagnostic;
use thiserror::Error;
pub type Result<T> = result::Result<T, Error>;

#[derive(Error, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Clone)]
pub enum Error {
    #[error("schema not exists {0:?}")]
    SchemaNotExists(String),
    #[error("schema already exists {0:?}")]
    SchemaAlreadyExists(String),
    #[error("graph type not exists {0:?}")]
    GraphTypeNotExists(String),
    #[error("graph type already exists {0:?}")]
    GraphTypeAlreadyExists(String),
    #[error("graph not exists {0:?}")]
    GraphNotExists(String),
    #[error("graph already exists {0:?}")]
    GraphAlreadyExists(String),
    #[error("procedure {0} not exists")]
    ProcedureNotExists(String),
    #[error("node type not exists {0:?}")]
    NodeTypeNotExists(String),
    #[error("edge type not exists {0:?}")]
    EdgeTypeNotExists(String),
    #[error("invalid field {0:?}")]
    InvalidField(String),
    #[error("duplicate field name {0:?}")]
    DuplicateField(String),
    #[error("not support operation {0:?}")]
    NotSupported(String),
    #[error("tmp error")]
    ErrorCur
}

// #[derive(Debug, Clone, PartialEq, Eq)]
// #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
// pub struct BinderError<T> {
//     input: Arc<str>,
//     span: Range<usize>,
//     position: (usize, usize),
//     #[cfg_attr(feature = "serde", serde(skip))]
//     _marker: PhantomData<T>,
// }

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NotSupportError {
    pub operation_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InvalidFieldError {
    pub field_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InvalidKeyError {
    pub key_name: String,
}



// impl<T> BinderError<T> {
//     pub fn input(&self) -> &Arc<str> {
//         &self.input
//     }
// 
//     pub fn span(&self) -> &Range<usize> {
//         &self.span
//     }
// 
//     pub fn position(&self) -> (usize, usize) {
//         self.position
//     }
// }

impl NotSupportError {
    pub fn operation_name(&self) -> &String {
        &self.operation_name
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SchemaNotExistsError {
    pub schema_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SchemaAlreadyExistsError {
    pub schema_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GraphTypeNotExistsError {
    pub type_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GraphTypeAlreadyExists {
    pub type_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GraphNotExistsError {
    pub graph_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GraphAlreadyExists {
    pub graph_name: String,
}

// impl_binder_not_exist_error_display!(
//     (SchemaNotExists, "schema"),
//     (GraphNotExists, "graph"),
//     (GraphTypeNotExists, "graph type"),
// );
//
// impl_binder_already_exist_error_display!(
//     (SchemaAlreadyExists, "schema"),
//     (GraphAlreadyExists, "graph"),
//     (GraphTypeAlreadyExists, "graph type"),
// );

impl Display for NotSupportError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let operation = &self.operation_name();
        write!(f,
            "{operation} is not support yet"
        )
    }
}
// 
// impl<T> core::error::Error for BinderError<T> where BinderError<T>: Debug + Display {}
impl core::error::Error for NotSupportError {}
// 
// #[cfg(feature = "miette")]
// impl<T> Diagnostic for BinderError<T> where BinderError<T>: core::error::Error {
//     fn source_code(&self) -> Option<&dyn miette::SourceCode> {
//         Some(&self.input)
//     }
// 
//     fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
//         Some(Box::new(core::iter::once(
//             miette::LabeledSpan::new_with_span(Some("here".into()), self.span.clone()),
//         )))
//     }
// }

#[cfg(feature = "miette")]
impl Diagnostic for NotSupportError {
    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
        Some(&self.operation_name)
    }
}

