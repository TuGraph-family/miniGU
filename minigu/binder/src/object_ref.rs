use std::ops::Deref;
use std::sync::Arc;

use minigu_catalog::provider::{
    DirectoryProvider, EdgeTypeProvider, GraphProvider, GraphTypeProvider, ProcedureProvider,
    SchemaProvider, VertexTypeProvider,
};
use serde::Serialize;
use smol_str::SmolStr;

pub type SchemaRef = ObjectRef<dyn SchemaProvider>;
pub type DirectoryRef = ObjectRef<dyn DirectoryProvider>;
pub type GraphRef = ObjectRef<dyn GraphProvider>;
pub type GraphTypeRef = ObjectRef<dyn GraphTypeProvider>;
pub type VertexTypeRef = ObjectRef<dyn VertexTypeProvider>;
pub type EdgeTypeRef = ObjectRef<dyn EdgeTypeProvider>;
pub type ProcedureRef = ObjectRef<dyn ProcedureProvider>;

#[derive(Debug, Serialize)]
pub struct ObjectRef<T: ?Sized> {
    pub name: SmolStr,
    #[serde(skip)]
    pub object: Arc<T>,
}

impl<T: ?Sized> ObjectRef<T> {
    pub fn new(name: SmolStr, object: Arc<T>) -> Self {
        Self { name, object }
    }

    pub fn name(&self) -> &SmolStr {
        &self.name
    }
}

impl<T: ?Sized> Deref for ObjectRef<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.object
    }
}
