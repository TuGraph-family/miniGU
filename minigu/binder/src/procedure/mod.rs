pub mod catalog;
pub mod object_ref;
pub mod procedure_call;
pub mod procedure_spec;
pub mod query;
pub mod type_element;
pub mod value_expr;
