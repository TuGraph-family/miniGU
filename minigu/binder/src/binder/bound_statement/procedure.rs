use gql_parser::ast::Yield;
use macro_rules_attribute::apply;

use crate::macros::base;
use crate::program::bound_statement::expr::BoundExpr;
use crate::program::bound_statement::object_ref::BoundProcedureRef;

#[apply(base)]
pub struct BoundCallProcedureStatement {
    pub optional: bool,
    pub procedure: BoundProcedureCall,
}

#[apply(base)]
pub enum BoundProcedureCall {
    Inline(BoundInlineProcedureCall),
    Named(BoundNamedProcedureCall),
}

#[apply(base)]
pub struct BoundInlineProcedureCall {}

#[apply(base)]
pub struct BoundNamedProcedureCall {
    pub name: BoundProcedureRef,
    pub args: Vec<BoundExpr>,
    // TODO: Bind Yield
    pub yield_clause: Option<Yield>,
}
