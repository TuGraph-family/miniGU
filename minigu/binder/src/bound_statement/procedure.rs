use gql_parser::ast::Yield;
use macro_rules_attribute::apply;
use serde::Serialize;
use crate::bound_statement::expr::BoundExpr;
use crate::bound_statement::object_ref::BoundProcedureRef;
use crate::catalog_ref::CallProcedureCatalogRef;

#[derive(Debug, Serialize)]
pub struct BoundCallProcedureStatement {
    pub optional: bool,
    pub procedure: BoundProcedureCall,
}

#[derive(Debug, Serialize)]
pub enum BoundProcedureCall {
    Inline(BoundInlineProcedureCall),
    Named(BoundNamedProcedureCall),
}

#[derive(Debug, Serialize)]
pub struct BoundInlineProcedureCall {}

#[derive(Debug, Serialize)]
pub struct BoundNamedProcedureCall {
    pub name: BoundProcedureRef,
    pub args: Vec<BoundExpr>,
    // TODO: Bind Yield
    pub yield_clause: Option<Yield>,
}
