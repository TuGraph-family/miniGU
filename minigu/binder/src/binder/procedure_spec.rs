use std::collections::HashMap;

use gql_parser::ast::{Procedure, Statement};
use itertools::Itertools;
use smol_str::SmolStr;

use super::Binder;
use crate::error::{BindResult, not_implemented};
use crate::procedure::procedure_spec::{BoundProcedure, BoundStatement};

impl Binder {
    pub fn bind_procedure(&mut self, procedure: &Procedure) -> BindResult<BoundProcedure> {
        if let Some(schema) = &procedure.at {
            let schema = self.bind_schema_ref(schema.value())?;
            self.current_schema = Some(schema);
        }
        if !procedure.binding_variable_defs.is_empty() {
            return not_implemented("binding variable definitions".to_string(), None);
        }
        let statement = self.bind_statement(procedure.statement.value())?;
        if !procedure.next_statements.is_empty() {
            return not_implemented("next statement".to_string(), None);
        }
        Ok(BoundProcedure {
            statement,
            next_statements: Vec::new(),
        })
    }

    pub fn bind_statement(&mut self, statement: &Statement) -> BindResult<BoundStatement> {
        match statement {
            Statement::Catalog(statements) => statements
                .iter()
                .map(|s| self.bind_catalog_modifying_statement(s.value()))
                .try_collect()
                .map(BoundStatement::Catalog),
            Statement::Query(statement) => {
                todo!()
            },
            Statement::Data(_) => {
                not_implemented("data-modifying statement".to_string(), None)
            }
        }
    }
}


