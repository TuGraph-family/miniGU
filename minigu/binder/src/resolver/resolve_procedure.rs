use std::sync::Arc;
use smol_str::ToSmolStr;
use gql_parser::ast::{CallProcedureStatement, NamedProcedureCall, ProcedureCall, ProcedureRef};
use minigu_catalog::provider::ProcedureProvider;
use crate::error::{BindError, BindResult};
use crate::program::Binder;
use crate::statement::procedure::{BoundCallProcedureStatement, BoundNamedProcedureCall, BoundProcedureCall};
use crate::statement::procedure_spec::BoundProcedure;
use crate::object_ref::ProcedureRef as ObjProcedureRef;

impl Binder {
    pub(crate) fn resolve_call_procedure(
        &mut self,
        call_procedure_statement: &CallProcedureStatement,
    ) -> BindResult<BoundCallProcedureStatement> {
        match call_procedure_statement.procedure.value() {
            ProcedureCall::Inline(call) => {
                return Err(BindError::NotSupported("Inline procedure".to_string()));
            }
            ProcedureCall::Named(named_procedure_call) => {
                let procedure_ref = named_procedure_call.name.value();
                let (schema_ref, procedure_name) = match procedure_ref {
                    ProcedureRef::Ref(procedure) => {
                        let schema_ref = match &procedure.schema {
                            Some(schema_ast) => self.resolve_schema_ref(schema_ast.value())?,
                            None => self.schema.as_ref().cloned().ok_or_else(||BindError::SchemaNotSpecified)?
                        };
                        let procedure_name = procedure.objects.iter().map(|ident| ident.value().to_string())
                            .collect::<Vec<_>>().join(".");
                        (schema_ref, procedure_name)
                    }
                    ProcedureRef::Parameter(param) => {
                        return Err(BindError::NotSupported("Procedure".to_string()));
                    }
                };

                let procedure_obj = schema_ref
                    .get_procedure(&procedure_name)
                    .map_err(|e| BindError::External(Box::new(e)))?
                    .ok_or_else(|| BindError::ProcedureNotFound(procedure_name.clone()))?;

                let bound_yield_index = self.resolve_yield_in_call_procedure(&procedure_obj, named_procedure_call)?;
                Ok(BoundCallProcedureStatement {
                    optional: call_procedure_statement.optional,
                    procedure: BoundProcedureCall::Named(
                        BoundNamedProcedureCall {
                            procedure_ref: ObjProcedureRef::new(
                                procedure_name.to_smolstr(),
                                procedure_obj),
                            yield_index: bound_yield_index,
                        }
                    )
                })
            }
        }
    }

    pub(crate) fn resolve_yield_in_call_procedure(
        &mut self,
        procedure_obj: &Arc<dyn ProcedureProvider>,
        call: &NamedProcedureCall,
    ) -> BindResult<Vec<usize>> {
        let yield_clause = call.yield_clause.clone();
        let schema_opt = procedure_obj.schema();
        let mut bound_yield_index = Vec::new();
        if let (Some(schema),Some(yield_clause)) = (schema_opt, yield_clause) {
            for yield_item in yield_clause.value() {
                let target_name = yield_item.value().name.value();
                let index_opt = schema.fields().iter().position(
                    |field| field.name() == target_name
                );
                match index_opt {
                    Some(index) => bound_yield_index.push(index),
                    None => {
                        return Err(BindError::NotSupported(format!(
                            "field `{}` not found in procedure fields",
                            target_name
                        )))
                    }
                }
            }
        }
        Ok(bound_yield_index)
    }
}


mod tests {
    use std::sync::Arc;
    use gql_parser::ast::ProgramActivity;
    use minigu_catalog::provider::CatalogRef;
    use crate::error::{BindError, BindResult};
    use crate::mock_catalog::MockCatalog;
    use crate::program::Binder;
    use crate::statement::procedure_spec::BoundProcedure;
    use insta::assert_yaml_snapshot;

    fn get_bound_procedure(gql: &str) -> BindResult<BoundProcedure> {
        let parsed = gql_parser::parse_gql(gql);
        let program_activity = parsed.unwrap().value().clone()
            .activity.unwrap().value().clone();
        let trans_activity = match program_activity {
            ProgramActivity::Session(session) => {
                return Err(BindError::NotSupported("Session".to_string()));
            }
            ProgramActivity::Transaction(transaction) => Some(transaction),
        };
        let procedure = trans_activity.unwrap().procedure.unwrap().value().clone();
        let catalog: CatalogRef = Arc::new(MockCatalog::default());
        let mut binder = Binder::new(catalog.clone(), None);
        binder.bind_procedure(&procedure)
    }
    
    #[test]
    fn test_resolve_procedure() {
        let stmt = get_bound_procedure("optional call /a/b.proc(1, 2, 3) yield a as a1, b as b1");
        assert!(matches!(stmt, Err(BindError::SchemaNotFound(_))));
        let stmt = get_bound_procedure("optional call /default/a/b/proc(1, 2, 3) yield a as a1, b as b1");
        assert!(matches!(stmt, Err(BindError::ProcedureNotFound(_))));
        let stmt = get_bound_procedure("optional call /default/a/b/procedure_test(1, 2, 3) yield a as a1, b as b1");
        let err_message = "field `a` not found in procedure fields".to_string();
        assert!(matches!(stmt, Err(BindError::NotSupported(err_message))));
        let stmt = get_bound_procedure("optional call /default/a/b/procedure_test(1, 2, 3) yield t1 as a1, t2 as b1");
        assert!(stmt.is_ok());
        assert_yaml_snapshot!(stmt.unwrap());
        let stmt = get_bound_procedure("at /default/a/b optional call procedure_test(1, 2, 3) yield t1 as a1");
        assert!(stmt.is_ok());
        assert_yaml_snapshot!(stmt.unwrap());
    }
}