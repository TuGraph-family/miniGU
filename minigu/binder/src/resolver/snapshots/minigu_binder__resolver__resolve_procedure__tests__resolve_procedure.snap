---
source: minigu/binder/src/resolver/resolve_procedure.rs
assertion_line: 125
expression: stmt.unwrap()
---
statement:
  Catalog:
    - Call:
        optional: true
        procedure:
          Named:
            procedure_ref:
              name: procedure_test
            yield_index:
              - 0
              - 1
next_statement: []
